src/main/kotlin/org/rust/cargo/runconfig/test/CargoTestRunConfigurationProducerBase.kt

```diff
@@ -8,98 +8,159 @@ package org.rust.cargo.runconfig.test
+        val elements = LangDataKeys.PSI_ELEMENT_ARRAY.getData(context.dataContext)?.toList()
+            ?: context.location?.psiElement?.let { listOf(it) }
+        return elements?.let { findTestConfig(it) }
+    }
+
+    private fun createConfigForDirectory(dir: PsiDirectory): TestConfig? {
+        val filesConfig = createConfigForMultipleFiles(dir.targets, climbUp = false) ?: return null
 
-        val targets = modulesWithBenches.mapNotNull { it.containingCargoTarget }
+        val sourceRoot = dir.sourceRoot ?: return null
+        val suitableSourceRootName = StringUtil.pluralize(commandName)
+        if (sourceRoot.name != suitableSourceRootName) return null
+
+        return DirectoryTestConfig(commandName, filesConfig.targets, dir)
+    }
+
+    private fun createConfigForMultipleFiles(elements: List<PsiElement>, climbUp: Boolean): TestConfig? {
+        val modConfigs = elements.mapNotNull { createConfigFor<RsMod>(listOf(it), climbUp) }
+
+        if (modConfigs.size == 1) {
+            return modConfigs.first()
+        }
+
+        val targets = modConfigs.flatMap { it.targets }
         if (targets.size <= 1) return null
 
         // If the selection spans more than one package, bail out.
-        val pkgs = targets.map { it.pkg }.distinct()
-        if (pkgs.size > 1) return null
+        val packages = targets.map { it.pkg }
+        if (packages.distinct().size > 1) return null
+
+        return MultipleFileTestConfig(commandName, targets, modConfigs.first().sourceElement)
+    }
+
+    private inline fun <reified T : RsQualifiedNamedElement> createConfigFor(
+        elements: List<PsiElement>,
+        climbUp: Boolean
+    ): TestConfig? {
+        val sourceElement = elements
+            .mapNotNull { findElement<T>(it, climbUp) }
+            .singleOrNull { isSuitable(it) && it.containingCargoTarget != null }
+            ?: return null
+        val target = sourceElement.containingCargoTarget ?: return null
+        val configPath = sourceElement.crateRelativePath.configPath() ?: return null
+        return SingleItemTestConfig(commandName, configPath, target, sourceElement)
+    }
 
-        return TestConfig.MultipleFileTestConfig(commandName, targets, modulesWithBenches[0])
+    private inline fun <reified T : PsiElement> findElement(base: PsiElement, climbUp: Boolean): T? {
+        if (base is T) return base
+        if (!climbUp) return null
+        return base.ancestorOrSelf()
     }
 
+    protected abstract fun isSuitable(element: PsiElement): Boolean
+
     companion object {
+        private val PsiDirectory.targets: List<RsFile>
+            get() {
+                val rootPath = virtualFile.pathAsPath
+                return project
+                    .cargoProjects
+                    .allProjects
+                    .asSequence()
+                    .mapNotNull { it.workspace }
+                    .flatMap { it.packages.asSequence() }
```

Comment:
We should process only workspace packages
Otherwise, we will provide run configuration for lib dependencies too
![image](https://user-images.githubusercontent.com/2539310/56431216-a6f0da00-62d1-11e9-8e3e-82789b3cd8a4.png)
And they will fail as expected
```
cargo test --color=always --no-run --package serde
error: package `serde` cannot be tested because it requires dev-dependencies and is not a member of the workspace
```

Reply:
Fixed (here and in `createConfigForCargoPackage`).